Open Problem: Preserve full users' privacy when providing and fetching Content (#5)
ipfs/research: OPEN_PROBLEMS/PRESERVE_USER_PRIVACY.md

# Preseve full users' privacy when providing and fetching Content

## Description

> The Web 2.0 and its centralized infrastructure missed to deliver full users' privacy when these are accessing and delivering content to other users in the Network. This can happen in one of two ways, either by simply transfering/storing the data uncrypted or by pattern analysis of access which can reveal users interest and intent.

Reviewer: ```suggestion
The Web 2.0 and its centralized infrastructure fail to protect users' privacy when they exchange data with other users. This can happen in one of two ways, either by simply transferring/storing the data unencrypted or by pattern analysis of access which can reveal interest and intent.
```
Question: is the concern re Web 2.0 eavesdropping by third parties or surveillance by intermediaries/data brokers? This and the next paragraphs seem to muddle the issue.

> The Web 2.0 and uses by default a centralized infrastructure design that fails to protect users' privacy. Some of the common patterns that we see making users vulnerable are: when data is not encrypted, both at rest and while transfering; users interactions with services leaks their intent, which creates the opportunity for a pattern analysis attack.

Reviewer: ```suggestion
The Web 2.0 implements a default centralized infrastructure design that fails to protect users' privacy. Some of the common patterns that we see making users vulnerable are: when data is not encrypted, both at rest and in transit, users' interactions with services leaks their intent, which creates the opportunity for a pattern analysis attack.
```

> In the Web 3.0, the dWeb, users get the ability to share data with other peers without using an intermediary, however, a complete solution is still missing that can prevent users leaking what data they are serving and fetching through sidechannels/pattern analysis (e.g. When searches are made, either through a search engine or simply by searching for the blocks in a Distributed Hash Table).

Reviewer: ```suggestion
In the Web 3.0, the dWeb, users get the ability to share data with other peers without using an intermediary. However, a complete solution is still missing that can prevent users leaking what data they are serving and fetching through side channels/pattern analysis (e.g. when searches are made, either through a search engine or simply by searching for the blocks in a Distributed Hash Table).
```

> Some solutions exist to mitigate this problem (see the State of the Art section below), however, non is yet complete as in "it is always 100% private, period", requiring users to adapt and adopt certain strategies to annonymize the content, depending on the type of interaction they have with other users. This Open Problem is beyond a data encryption or wire encryption problem, a complete solution will have to provide a way to give access to others to access data and later revoke (Authorization), provide a users with the ability to know when a piece of data they have share is being accessed (Accounting) and offer a guarantee that the data is being shared with exactly the right user and no one else (Authentication), this while not letting a third party understand what is being shared, how and with whom.

Reviewer: ```suggestion
Some solutions exist to mitigate this problem (see the State of the Art section below). However, none is yet complete as in "it is always 100% private, period", requiring users to adapt and adopt certain strategies to anonymize the content, depending on the type of interaction they have with other users. This Open Problem is beyond a data encryption or wire encryption problem; a complete solution will have to provide a way to give access to others to access data and later revoke (Authorization), provide users with the ability to know when a piece of data they have shared is being accessed (Accounting) and offer a guarantee that the data is being shared with exactly the right user and no one else (Authentication), this while not letting a third party understand what is being shared, how and with whom.
```

> Some solutions exist to mitigate this problem (see the State of the Art section below). However, none is yet complete as in "it is always 100% private, period", requiring users to adapt and adopt certain strategies to anonymize the content, depending on the type of interaction they have with other users. This Open Problem is beyond a data encryption or wire encryption problem; a complete solution will have to provide a way to give access to others to access data and later revoke (Authorization), provide users with the ability to know when a piece of data they have shared is being accessed (Accounting) and offer a guarantee that the data is being shared with exactly the right user and no one else (Authentication), this while not letting a third party understand what is being shared, how and with whom.

Reviewer: ```suggestion
Some solutions exist to mitigate this problem (see the State of the Art section below). However, none is yet complete as in "it is always 100% private, period", requiring users to adapt and adopt certain strategies to anonymize the content, depending on the type of interaction they have with other users. This Open Problem is beyond a data encryption or wire encryption problem; a complete solution will have to provide a way to grant and revoke other users' access to content (Authorization), provide users with the ability to know when a piece of content they have shared is being accessed, (Accounting) and guarantee that the users with whom the content is being shared are who they claim to be (Authentication), while not letting a third party understand what is being shared, how and with whom.
```
The description of authentication was reading too much like authorisation. Replaced data with content just to prevent the usual "data is plural" issue.

This happens to be one of the toughest problems to solve in order to provide a complete and human rights preserving fabric for knowledge.

## State of the Art

> This survey on the State of the Art is not by any means complete, however, it should provide a good entry point to learn what are the existing work. If you have something that is fundamentally missing, please consider submitting a PR to augment this survey. 

### Within the IPFS Ecosystem
> Existing attempts and strategies

##### Wire Encryption

Thanks to libp2p, IPFS ensures that the communication between any two IPFS nodes is always encrypted. This is achieved through one of the many Crypto Channels that libp2p supports such as: SECIO, TLS 1.3, DTLS (in WebRTC) and NOISE.

##### Data Encryption

> Data Encryption means that we encrypt data at rest and only decrypted to consume it. This means that only the owner of the decryption key can access it. Some solutions with this technique are:

Reviewer: ```suggestion
Data Encryption means that we encrypt data at rest and only decrypt it to consume it. This means that only the owner of the decryption key can access it. Some solutions using this technique are:
```
Is this an accepted definition of data encryption? I feel like that's a much broader concept.

Author: Good point. I'll reword.

> - [ipfs-senc](https://github.com/jbenet/ipfs-senc) - Encrypts the data with a symetrical key that is shared to the receiver through a sidechannel

Reviewer: ```suggestion
- [ipfs-senc](https://github.com/jbenet/ipfs-senc) - Encrypts the data with a symmetric key that is shared to the receiver through a sidechannel
```

##### Capability Systems / Cryptographic ACLs

> - [peer-base cryptographic ACLs](https://github.com/peer-base/peer-base) - These are used by [PeerPad](https://peerpad.net). The way they work is that for each user, a Pub/Priv key pair is generated and every time a user wants to make a modification, signs that modification and then encrypts it with a symmetric room key, so that only owners of the symmetric key can change and only changes from valid peers are accepted.

Reviewer: ```suggestion
- [peer-base cryptographic ACLs](https://github.com/peer-base/peer-base) - These are used by [PeerPad](https://peerpad.net). For each user, a Public/Private key pair is generated. Every time a user wants to make a modification, the user signs that modification and encrypts it with a symmetric room key so that only owners of the symmetric key can change and only changes from valid peers are accepted.
```

> - [peer-base cryptographic ACLs](https://github.com/peer-base/peer-base) - These are used by [PeerPad](https://peerpad.net). For each user, a Public/Private key pair is generated. Every time a user wants to make a modification, the user signs that modification and encrypts it with a symmetric room key so that only owners of the symmetric key can change and only changes from valid peers are accepted.

Author: - [ ] link to video Demo from @pgte in showcasing PeerPad ACL

Author: Also related, include https://www.youtube.com/watch?v=yc10xRifTmQ&feature=youtu.be

Author: @pgte can you help me find your video?

##### Private/Disjoint Networks

> Creating a separate IPFS Network will ensure that only the nodes can access the content within that network. 

Reviewer: ```suggestion
Creating a separate IPFS Network will ensure that only member nodes can access the content within that network. 
```

> - [libp2p-pnet](https://github.com/libp2p/specs/blob/master/pnet/Private-Networks-PSK-V1.md) takes that one step forward and creates a protection using a pre-shared key. This means that only the owners of that key can join this network (to prevent from mistakenly joining two networks and making all data accessible).

Author: - [ ] Add video demo that @jacobheun recorded

Author: @jacobheun I'm a terrible person and somehow I forgot to add the url when you sent it to me and now I can't find it again :( Could you send it to me again 🙏🏽 thank you!

- [Using Sphinx to Improve Onion RoutingCircuit Construction](https://www.cypherpunks.ca/~iang/pubs/SphinxOR.pdf)
- [Scuttlebutt Secret Handshare](https://dominictarr.github.io/secret-handshake-paper/shs.pdf)
- [Octopus: A Secure and Anonymous DHT Lookup](https://ieeexplore.ieee.org/document/6258005)
- [Vuvuzela: Scalable Private Messaging Resistant to Traffic Analysis](https://davidlazar.org/papers/vuvuzela.pdf)
- [Define privacy threat model for DHTs (and other overlay P2P networks)](https://github.com/gpestana/notes/issues/3)
- [Talek: a Private Publish-Subscribe Protocol](https://raymondcheng.net/download/papers/talek-tr.pdf)
- [SoK: Secure Messaging](https://ieeexplore.ieee.org/document/7163029)
- [Ricochet](https://github.com/ricochet-im/ricochet/blob/master/doc/protocol.md)
- [Cwtch: Privacy Preserving Infrastructure for Asynchronous,Decentralized, Multi-Party and Metadata Resistant Applications](https://cwtch.im/cwtch.pdf)
- [HOPR - privacy-preserving messaging protocol ](https://github.com/validitylabs/hopr)

### Known shortcommins of existing solutions
> What are the limitations on those solutions?

Some of the known shortcommings of existing solutions are:

> - They don't offer protection to network analysis (it is possible to infer what the user is doing by analysis the network traffic)

Reviewer: ```suggestion
- They don't offer protection against network analysis (it is possible to infer what the user is doing by analysing network traffic)
```

- Solutions that are more resistant (not fully resistent) typically trade off bandwidth + memory for creating that protection (e.g. creating noise in the network to make it hard to distinguish valid from dummy traffic) 
- Lack of data encryption at rest
- Lack of complete authorization + revocation

## Solving this Open Problem

### What is the impact

> Finding a complete solution to this Open Problem will grant the ability for any human in the planet to privately share information, read content without leaking their intent to anyone else other than the provider of that content, or even possibly, not even to the provider.

Reviewer: ```suggestion
Finding a complete solution to this Open Problem will grant any human in the planet the ability to privately access or share information without leaking their intent to non-local observers other than the provider of that content. Solutions may also be able to hide said intent from the content provider.
```

It will lead to a generation of safer applications, both consumer level and business critical, that require full control of one's data (e.g. Health Data).

### What defines a complete solution?
> What hard constraints should it obey? Are there additional soft constraints that a solution would ideally obey?

Valid solutions should improve the current state of the art or offer definitive solutions for:

> - [ ] Two users can transfer a piece of information that no other user knows or can guess: what it is, why it is being transfer, between whom and when.

Reviewer: ```suggestion
- [ ] Two users can transfer a piece of information without any other user knowing or being able to predict: what it is, why it is being transferred, when and between whom.
```

> - [ ] No single central authorities are required to mediate the communication

Reviewer: This is not incorrect but **single** ... authoriti**es** always read a bit funny.

Author: ```suggestion
- [ ] No single central authority are required to mediate the communication
```

##### Private/Disjoint Networks

Creating a separate IPFS Network will ensure that only the nodes can access the content within that network. 

- [libp2p-pnet](https://github.com/libp2p/specs/blob/master/pnet/Private-Networks-PSK-V1.md) takes that one step forward and creates a protection using a pre-shared key. This means that only the owners of that key can join this network (to prevent from mistakenly joining two networks and making all data accessible).

##### Onion Routing 

Onion routing is a technique for anonymous communication over a computer network. In an onion network, messages are encapsulated in layers of encryption, analogous to layers of an onion. The encrypted data is transmitted through a series of network nodes called onion routers, each of which "peels" away a single layer, uncovering the data's next destination.

- [libp2p-onion](https://github.com/OpenBazaar/go-onion-transport)

> ### Within the broad Research Ecosystem

Author: Note: Once this Open Problem statement is reviewed and merged, I see part as my next action items is to go through this literature and start adding appendix with notes from all this projects. The goal being to make all this research shared knowledge by the team by making it easier to consume.

Reviewer: I'm sure @yiannisbot will also have things to add to the list and, potentially, detailed knowledge of some of them, making it unnecessary to go through those in detail.

Author: Ping @yiannisbot

##### Data Encryption

By encrypting the data at rest and only decrypting it when it needs to be used, we can ensure that only someone with access to the decryption key can indeed access that data. Some solutions using this technique are:

- [ipfs-senc](https://github.com/jbenet/ipfs-senc) - Encrypts the data with a symmetric key that is shared to the receiver through a sidechannel

##### Capability Systems / Cryptographic ACLs

- [peer-base cryptographic ACLs](https://github.com/peer-base/peer-base) - These are used by [PeerPad](https://peerpad.net). For each user, a Public/Private key pair is generated. Every time a user wants to make a modification, the user signs that modification and encrypts it with a symmetric room key so that only owners of the symmetric key can change and only changes from valid peers are accepted.

##### Private/Disjoint Networks

Creating a separate IPFS Network will ensure that only member nodes can access the content within that network. 

> - [libp2p-pnet](https://github.com/libp2p/specs/blob/master/pnet/Private-Networks-PSK-V1.md) takes that one step forward and creates a protection using a pre-shared key. This means that only the owners of that key can join this network (to prevent from mistakenly joining two networks and making all data accessible).

Author: ```suggestion
- [libp2p-pnet](https://github.com/libp2p/specs/blob/master/pnet/Private-Networks-PSK-V1.md) takes that one step forward and creates a protection using a pre-shared key. This means that only the owners of that key can join this network (to prevent from mistakenly joining two networks and making all data accessible).
  - See a demo of it at https://youtu.be/fObld4alGag?t=47
```

> # Preseve full users' privacy when providing and fetching Content

Reviewer: ```suggestion
# Preserve user privacy when providing and fetching content
```
There's a typo in the doc's first word 😄 The rest is mostly aesthetics.

## Short Description
> In one sentence or paragraph.

> How to ensure that the user's of the IPFS network can collect and provide information while mainting their full anonymity.

Reviewer: ```suggestion
How to ensure that users of the IPFS network can retrieve and provide information while maintaining full anonymity.
```

Author: Thank you @jsoares! You are a saint for reviewing my texts with so much detail, really appreciate it!

## Description

The Web 2.0 implements a default centralized infrastructure design that fails to protect users' privacy. Some of the common patterns that we see making users vulnerable are: when data is not encrypted, both at rest and in transit, users' interactions with services leaks their intent, which creates the opportunity for a pattern analysis attack.

In the Web 3.0, the dWeb, users get the ability to share data with other peers without using an intermediary. However, a complete solution is still missing that can prevent users leaking what data they are serving and fetching through side channels/pattern analysis (e.g. when searches are made, either through a search engine or simply by searching for the blocks in a Distributed Hash Table).

> Some solutions exist to mitigate this problem (see the State of the Art section below). However, none is yet complete as in "it is always 100% private, period", requiring users to adapt and adopt certain strategies to anonymize the content, depending on the type of interaction they have with other users. This Open Problem is beyond a data encryption or wire encryption problem; a complete solution will have to provide a way to grant and revoke other users' access to content (Authorization), provide users with the ability to know when a piece of content they have shared is being accessed, (Accounting) and guarantee that the users with whom the content is being shared are who they claim to be (Authentication), while not letting a third party understand what is being shared, how and with whom.

Reviewer 2: I agree with this paragraph, specially as a motivation. I think it would be interesting to mention that different applications and use cases have different needs and threat models. Providing a 100% private system is very hard -- specially across the whole stack -- and "privacy" may mean different things depending on the context. Setting the goal as 100% privacy feels a bit unrealistic and non practical.

Reviewer: Some random thoughts:
- 100% privacy is suggestive of anonymity.
- Content verification does not solve every issue. If there's a malicious requester spamming the network or a malicious provider feeding corrupted data, this will not lead to successful content forgery but can lead to DoS. Authentication can be used to stop bad actors.
- Authorisation/revocation usually depends on authentication. It does not necessarily reveal information to the network, though; could be offline knowledge of a private key. Authentication can preserve anonymity. 
- Accounting, depending on how it is defined, may require verified identities. 

I think this goes back to both your points that not everything has to be done at the same layer, though.

##### Data Encryption

By encrypting the data at rest and only decrypting it when it needs to be used, we can ensure that only someone with access to the decryption key can indeed access that data. Some solutions using this technique are:

- [ipfs-senc](https://github.com/jbenet/ipfs-senc) - Encrypts the data with a symmetric key that is shared to the receiver through a sidechannel

##### Capability Systems / Cryptographic ACLs

- [peer-base cryptographic ACLs](https://github.com/peer-base/peer-base) - These are used by [PeerPad](https://peerpad.net). For each user, a Public/Private key pair is generated. Every time a user wants to make a modification, the user signs that modification and encrypts it with a symmetric room key so that only owners of the symmetric key can change and only changes from valid peers are accepted.

##### Private/Disjoint Networks

Creating a separate IPFS Network will ensure that only member nodes can access the content within that network. 

> 

Reviewer 2: New paper published few days ago on encrypted DHTs:

- Encrypted Distributed Hash Tables (https://eprint.iacr.org/2019/1126)

- [libp2p-pnet](https://github.com/libp2p/specs/blob/master/pnet/Private-Networks-PSK-V1.md) takes that one step forward and creates a protection using a pre-shared key. This means that only the owners of that key can join this network (to prevent from mistakenly joining two networks and making all data accessible).
  - See a demo of it at https://youtu.be/fObld4alGag?t=47

##### Onion Routing 

Onion routing is a technique for anonymous communication over a computer network. In an onion network, messages are encapsulated in layers of encryption, analogous to layers of an onion. The encrypted data is transmitted through a series of network nodes called onion routers, each of which "peels" away a single layer, uncovering the data's next destination.

- [libp2p-onion](https://github.com/OpenBazaar/go-onion-transport)

> 

Reviewer 2: I'd suggest also adding [p3lib-sphinx](https://github.com/hashmatter/p3lib/tree/master/sphinx) here, which implements the sphinx onion packet format along with logic for packet construction and relaying. This protocol is used in lightning network and in Nym and  there is also [a demo of p3lib-sphinx working with libp2p](https://github.com/hashmatter/libp2p-onion-routing), [and a video](https://www.youtube.com/watch?v=j64C5CTb8J8). Basically it allows libp2p nodes to delegate DHT requests using other libp2p nodes as relayers. This is pretty cool since you don't have to plug in/depend on any other overlay (i2p/Tor) and provide similar unlikeability and privacy properties (providing that you have enough entropy in the network and do not leak the circuit path).

Reviewer 2: Actually I think this is an interesting research topic: what are the advantages and disadvantages of onion routing (and mixnet) using only libp2p nodes ("onion routing on top of the DHT overlay"), versus forwarding the traffic through e.g. Tor/Nym. We've been thinking about this and there's a lot to consider: e.g. how are relayer nodes privately discovered and selected (leaking nodes in the secure path is a massive vulnerability)? entropy/noise? etc

Another way to phrase this is: imagine a [libp2p circuit relay](https://github.com/libp2p/specs/tree/master/relay) with "privacy preserving properties": the circuit would be consist of 3 relay nodes where the relayers in the path would not be able to learn its position in the circuit (so as not being able to identify the ID of the DHT initiator).  

(more notes/ideas on this topic https://github.com/gpestana/notes/issues/19)

Author: Added your notes, thank you!

### Within the broad Research Ecosystem
> How do people try to solve this problem?

- Authorization, Authentication, Accounting
  - [TahoeLAFS Capability System](https://en.wikipedia.org/wiki/Tahoe-LAFS) - The peer-base Cryptographic ACLs were inspired by TahoeLAFS Capabolity System. 
- Data Encryption
  - [DoubleRatchet](https://signal.org/docs/specifications/doubleratchet/) / [Matrix.org Olm/MegaOLM](https://gitlab.matrix.org/matrix-org/olm/blob/master/docs/olm.md#olm-a-cryptographic-ratchet)
- M2M Private Communication
  - [Scuttlebutt Secret Handshare](https://dominictarr.github.io/secret-handshake-paper/shs.pdf)
  - [HOPR - privacy-preserving messaging protocol ](https://github.com/validitylabs/hopr)
  - [SoK: Secure Messaging](https://ieeexplore.ieee.org/document/7163029)
  - [Talek: a Private Publish-Subscribe Protocol](https://raymondcheng.net/download/papers/talek-tr.pdf)
  - [Ricochet](https://github.com/ricochet-im/ricochet/blob/master/doc/protocol.md)

> - Content Routing

Reviewer 2: - ShadowWalker: peer-to-peer anonymous communication using redundant structured topologies (https://dl.acm.org/citation.cfm?id=1653683)

  - [Define privacy threat model for DHTs (and other overlay P2P networks)](https://github.com/gpestana/notes/issues/3)
  - [Using Sphinx to Improve Onion Routing Circuit Construction](https://www.cypherpunks.ca/~iang/pubs/SphinxOR.pdf)
  - [Octopus: A Secure and Anonymous DHT Lookup](https://ieeexplore.ieee.org/document/6258005)
- Traffic Analysis resistant
  - [Vuvuzela: Scalable Private Messaging Resistant to Traffic Analysis](https://davidlazar.org/papers/vuvuzela.pdf)
- Other
  - [Cwtch: Privacy Preserving Infrastructure for Asynchronous,Decentralized, Multi-Party and Metadata Resistant Applications](https://cwtch.im/cwtch.pdf)

> 

Reviewer 2: I think Private Information Retrieval (PIR) systems will be a very useful building block for privacy in P2P networks, so I suggest:

- [XPIR : Private Information Retrieval for Everyone](https://eprint.iacr.org/2014/1025.pdf)
- [PIR with compressed queries and amortized query processing](https://eprint.iacr.org/2017/1142.pdf)

And the [Sinkhole protocol](https://github.com/gpestana/notes/issues/27) which relies on computational PIR for enhancing privacy of DHT lookups. (specs still being fleshed out)

### Known shortcommins of existing solutions
> What are the limitations on those solutions?

Some of the known shortcommings of existing solutions are:

- They don't offer protection against network analysis (it is possible to infer what the user is doing by analysing network traffic)
- Solutions that are more resistant (not fully resistent) typically trade off bandwidth + memory for creating that protection (e.g. creating noise in the network to make it hard to distinguish valid from dummy traffic) 
- Lack of data encryption at rest
- Lack of complete authorization + revocation

> 

Reviewer 2: - Some of the solutions (e.g. OctopusDHT) rely on centralised certificate authorities for reputation management

- How to measure privacy? (old but gold)

## Solving this Open Problem

### What is the impact

Finding a complete solution to this Open Problem will grant any human in the planet the ability to privately access or share information without leaking their intent to non-local observers other than the provider of that content. Solutions may also be able to hide said intent from the content provider.

It will lead to a generation of safer applications, both consumer level and business critical, that require full control of one's data (e.g. Health Data).

### What defines a complete solution?
> What hard constraints should it obey? Are there additional soft constraints that a solution would ideally obey?

Valid solutions should improve the current state of the art or offer definitive solutions for:

- [ ] Two users can transfer a piece of information without any other user knowing or being able to predict: what it is, why it is being transferred, when and between whom.
- [ ] No single central authority are required to mediate the communication
- [ ] A provider has a way to grant and revoke access to information.

> 

Reviewer 2: I would add:

- A user can *discover* public information without leaking information about her interest. An adversary should not be able to link two or more discovery requests either.

This idea encapsulates the problem with DHTs. I want X but I rely on the network to find X. Thus, I need to disclose to peers I'm interested about X. Huge tension between practicality and privacy.

Some solutions exist to mitigate this problem (see the State of the Art section below). However, none is yet complete as in "it is always 100% private, period", requiring users to adapt and adopt certain strategies to anonymize the content, depending on the type of interaction they have with other users. This Open Problem is beyond a data encryption or wire encryption problem; a complete solution will have to provide a way to grant and revoke other users' access to content (Authorization), provide users with the ability to know when a piece of content they have published is being accessed (Accounting), and guarantee that the users with whom the content is being shared are who they claim to be (Authentication), while not letting a third party understand what is being shared, how and with whom.

This happens to be one of the toughest problems to solve in order to provide a complete and human rights preserving fabric for knowledge.

> ## Background

Author: ```suggestion
### Background
```

##### Private/Disjoint Networks

Creating a separate IPFS Network will ensure that only member nodes can access the content within that network. 

- [libp2p-pnet](https://github.com/libp2p/specs/blob/master/pnet/Private-Networks-PSK-V1.md) takes that one step forward and creates a protection using a pre-shared key. This means that only the owners of that key can join this network (to prevent from mistakenly joining two networks and making all data accessible).
  - See a demo of it at https://youtu.be/fObld4alGag?t=47

##### Onion Routing 

Onion routing is a technique for anonymous communication over a computer network. In an onion network, messages are encapsulated in layers of encryption, analogous to layers of an onion. The encrypted data is transmitted through a series of network nodes called onion routers, each of which "peels" away a single layer, uncovering the data's next destination.

- [libp2p-onion](https://github.com/OpenBazaar/go-onion-transport)

### Within the broad Research Ecosystem
> How do people try to solve this problem?

> A general and rather simple approach and rule of thumb is to use an anonymising proxy which is placed between a sender and a receiver of traffic. This inevitably means that there is trust placed on this proxy to behave as desired. Decentralised approaches (that don't necessarily have to trust proxies or proxy operators) rely on layered encryption and multiple extra hops of routing so that any message cannot be linked to its originator/source.

Author: @yiannisbot (https://github.com/ipfs/research/pull/10#discussion_r328521147) I feel that calling this _general_ might lead to misunderstandings. How is this general? Which systems use this approach?
